ARROW-16595: [WEBSITE] DataFusion 8.0.0 release blog post (#217)
apache/arrow-site: _posts/2022-05-16-datafusion-8.0.0.md

contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

# Introduction

[DataFusion](https://arrow.apache.org/datafusion/) is an extensible query execution framework, written in Rust, that
uses Apache Arrow as its in-memory format.

When you want to extend your Rust project with [SQL support](https://arrow.apache.org/datafusion/user-guide/sql/sql_status.html),
a DataFrame API, or the ability to read and process Parquet, JSON, Avro or CSV data, DataFusion is definitely worth
checking out.

> DataFusion's  SQL, `DataFrame`, and manual `PlanBuilder` API let users access a sophisticated query optimizer and

Reviewer: ```suggestion
DataFusion's SQL, `DataFrame`, and manual `PlanBuilder` API let users access a sophisticated query optimizer and
```

execution engine capable of fast, resource efficient, and parallel execution that takes optimal advantage of

> today's multicore hardware. Being written in Rust means DataFusion can offer *both* the safety of dynamic languages and

Reviewer: ```suggestion
today's multicore hardware. Being written in Rust means DataFusion can offer *both* the safety of a dynamic language and
```

- `CREATE EXTERNAL TABLE` now supports JSON files, `IF NOT EXISTS`, and partition columns

## SQL Support

The SQL query planner now supports a number of new SQL features, including:

- _Subqueries_: when used via `IN`, `EXISTS`, and as scalars
- _Grouping Sets_: `CUBE` and `ROLLUP` grouping sets.
- _Aggregate functions_: `approx_percentile`, `approx_percentile_cont`, `approx_percentile_cont_with_weight`, `approx_distinct`, `approx_median` and `array`
- _`null` literals_
- _bitwise operations_: for example '|'

There are also many bug fixes and improvements around normalizing identifiers consistently.

We continue our tradition of incrementally releasing support for new
features as it is developed. Thus, while the physical plan may not yet
support all new features, it gets more complete each release. These
changes also make DataFusion an increasingly compelling choice for
projects looking for a SQL parser and query planner to translate to
their own execution engine.

## Query Execution & Internals

There are several notable improvements and new features in the query execution engine:

- The `ExecutionContext` has been renamed to `SessionContext` and now supports multi-tenancy

> - The `ExecutionPlan` trait is no longer `async`

Reviewer: 👍

- A new serialization API for serializing plans to bytes (based on protobuf)

In addition, we have added several foundational features to drive even

> more advanced query processing into DataFusion, focused on running

Reviewer: ```suggestion
more advanced query processing into DataFusion, focusing on running
```

arbitrary queries larger than available memory, and pushing the

> envelope for performance of sorting, grouping and joining even

Reviewer: ```suggestion
envelope for performance of sorting, grouping, and joining even
```

further:

- Morsel-Driven Scheduler based on ["Morsel-Driven Parallelism: A NUMA-Aware Query
  Evaluation Framework for the Many-Core Age"](https://15721.courses.cs.cmu.edu/spring2016/papers/p743-leis.pdf)
- Consolidated object store implementation and integration with parquet decoding

> - Memory Limited Spilling Sort operator

Reviewer: ```suggestion
- Memory Limited Spilling sort operator
```

to the scheduler to support UDFs, and there are some robustness improvements, such as cleaning up work directories
and persisting session configs to allow schedulers to restart and continue processing in-flight jobs.

## Upcoming Work

Here are some of the initiatives that the community plans on working on prior to the next release.

- There is a [proposal to move Ballista to its own top-level arrow-ballista repository](https://docs.google.com/document/d/1jNRbadyStSrV5kifwn0khufAwq6OnzGczG4z8oTQJP4/edit?usp=sharing)
 to decouple DataFusion and Ballista releases and to allow each project to have documentation better targeted at
  its particular audience.
- We plan on increasing the frequency of DataFusion releases, with monthly releases now instead of quarterly. This
  is driven by requests from the increasing number of projects that now depend on DataFusion.
- There is ongoing work to implement new optimizer rules to rewrite queries containing subquery expressions as
  joins, to support a wider range of queries.
- The new scheduler based on morsel-driven execution will continue to evolve in this next release, with work to

>  refine IO abstractions to improve performance and integration with the new scheduler.

Reviewer: ```suggestion
  refine IO abstractions to improve performance and integration with the new scheduler.
```
